Provide ML completion ranking for Rust

ortem opened the pull request:
Part of https://github.com/intellij-rust/intellij-rust/issues/6411

This PR adds a new module `ml-completion` to the project. This module provides the ML-based code completion ranking support for Rust. Currently, it consists of:
* `RsCompletionFeaturesPolicy` enabling Ngram model for future models
* `RsElementFeatureProvider` providing Rust-specific features to train future models, with tests
* `RsMLRankingProvider` providing an initial ranking model (trained without the stuff above, therefore not very intelligent so far)

Thanks to @bibaev for help!

### How to use
The new ranking model can be enabled in `Editor | General | Code Completion | Machine Learning-Assisted Completion` settings (see [help](https://www.jetbrains.com/help/idea/auto-completing-code.html#ml_completion) for more details). However, at the moment, the model is very rough, so it's not really recommended for enabling. 

<img width="400" alt="settings" src="https://user-images.githubusercontent.com/4854600/99807852-1f281d00-2b51-11eb-97d1-02ffa11f024f.png">

<img width="400" alt="before" src="https://user-images.githubusercontent.com/4854600/99807849-1df6f000-2b51-11eb-8e25-61107f966d18.png">

<img width="400" alt="after" src="https://user-images.githubusercontent.com/4854600/99807843-1b949600-2b51-11eb-9c7c-17fbaff722b3.png">

bibaev commented on build.gradle.kts:
```diff
@@ -45,6 +45,8 @@ val javaPlugin = "java"
 // BACKCOMPAT: 2020.2
 val clionPlugins = if (platformVersion < 203) emptyList() else listOf("com.intellij.cidr.base", "com.intellij.clion")
+val mlCompletionPlugin = "com.intellij.stats.completion"
+val mlCompletionRankingRustPlugin = "org.jetbrains.intellij.deps.completion:completion-ranking-rust:0.0.4"
```
I'd suggest renaming the value to "mlCompletionModelLibrary"

bibaev commented on ml-completion/src/main/kotlin/RustElementFeatureProvider.kt:
```diff
@@ -0,0 +1,197 @@
+        element: LookupElement,
+        location: CompletionLocation,
+        contextFeatures: ContextFeatures
+    ): MutableMap<String, MLFeatureValue> {
```
This method might return immutable `Map<String, MLFeatureValue>`

bibaev commented on ml-completion/src/main/kotlin/RustMLRankingProvider.kt:
```diff
@@ -0,0 +1,15 @@
+import com.intellij.lang.Language
+
+@Suppress("UnstableApiUsage")
+class RustMLRankingProvider : CatBoostJarCompletionModelProvider("Rust", "rust_features", "rust_model") {
```
It worth adding a test to check that metadata we ship with the model is fully consistent. See example here -  `com.jetbrains.completion.ml.ranker.MetadataConsistencyTest#testRustMetadata`

Undin commented on ml-completion/src/main/kotlin/RustElementFeatureProvider.kt:
```diff
@@ -0,0 +1,197 @@
+}
+
+/** See [org.rust.lang.core.psi.RsTokenType] */
+internal enum class RsKeywordMLKind {
```
Let's somehow check that this list contains exactly the same elements as `org.rust.lang.core.psi.RsTokenTypeKt#RS_KEYWORDS` not to forgot to add the corresponding element to the second list when you add a new keyword element

ortem replied:
Added `test all keywords are covered` test to `RsElementFeatureProviderTest`

ortem commented on ml-completion/src/main/kotlin/RustCompletionFeaturesPolicy.kt:
```diff
@@ -0,0 +1,12 @@
+
+import com.intellij.completion.ml.features.CompletionFeaturesPolicy
+
+class RustCompletionFeaturesPolicy : CompletionFeaturesPolicy {
```
Renamed, as well as `RustElementFeatureProvider` -> `RsElementFeatureProvider`

ortem commented on ml-completion/src/main/kotlin/RustElementFeatureProvider.kt:
```diff
@@ -0,0 +1,197 @@
+        val rsElement = psiElement as? RsElement ?: return result
+        val rsPsiElementKind = RsPsiElementMLKind.from(rsElement)
+        if (rsPsiElementKind != null) {
+            result["kind"] = MLFeatureValue.categorical(rsPsiElementKind)
```
Extracted to
```Kotlin
private object Features {
    const val Kind: String = "kind"
    const val IsFromStdlib: String = "is_from_stdlib"
}
```

ortem commented on ml-completion/src/test/kotlin/RustModelMetadataConsistencyTest.kt:
```diff
@@ -0,0 +1,12 @@
+
+class RustModelMetadataConsistencyTest {
+    @Test
+    fun testMetadata() = RustMLRankingProvider().assertModelMetadataConsistent()
```
Changed to `test model metadata consistency`

Undin reviewed (approved):
LGTM except minor comments

ortem commented on ml-completion/src/main/kotlin/org/rust/ml/RustMLRankingProvider.kt:
```diff
@@ -0,0 +1,15 @@
+import org.rust.lang.RsLanguage
+
+@Suppress("UnstableApiUsage")
+class RustMLRankingProvider : CatBoostJarCompletionModelProvider("Rust", "rust_features", "rust_model") {
```
I've chosen the name with "Rust" word because all the `CatBoostJarCompletionModelProvider` subclasses use the full language names instead of the abbreviations (including our experimental model provider for EAP builds `ExperimentRustMLRankingProvider`). So it would be easier to search across all the implementations with `Find external Usages` action for those who unfamiliar with our "Rs" prefix, for example

Undin reviewed (changes requested):
Could you rebase into master, please? It will show the problem with `com.intellij.stats.completion` dependency (CI faces it because it actually checks master +  your changes)

Undin commented on ml-completion/src/test/resources/META-INF/plugin.xml:
```diff
@@ -0,0 +1,5 @@
+<idea-plugin xmlns:xi="http://www.w3.org/2001/XInclude" allow-bundled-update="true">
+    <id>org.rust.ml</id>
+    <xi:include href="/META-INF/rust-core.xml" xpointer="xpointer(/idea-plugin/*)"/>
+    <depends optional="true" config-file="ml-completion-only.xml">com.intellij.stats.completion</depends>
```
plugin id is not updated to `com.intellij.completion.ml.ranking`, as a result, test on 2020.3 fail

Undin commented on build.gradle.kts:
```diff
@@ -45,6 +45,7 @@ val javaPlugin = "java"
 val javaScriptPlugin = "JavaScript"
 // BACKCOMPAT: 2020.2
 val clionPlugins = if (platformVersion < 203) emptyList() else listOf("com.intellij.cidr.base", "com.intellij.clion")
+val mlCompletionPlugin = "com.intellij.completion.ml.ranking"
```
Looks like `com.intellij.completion.ml.ranking` was introduced only in 2020.3 (extracted from `com.intellij.stats.completion`) so current impls cannot be applied to 2020.2

I see 2 variants:
- try to support 2020.2 as well. It leads to a lot of conditions in build scripts and xmls
- support only 2020.3

I prefer the second option (support only 2020.3) because it's simpler (for example, `CatBoostJarCompletionModelProvider` exists only in 2020.3) and we are going to drop 2020.2 support quite soon

ortem replied:
I agree with you, so I moved all the code into `203` folder

Undin reviewed (commented):
Also, there is a merge conflict

Undin reviewed (commented):
Note, I've added missed keywords in #6492 (it was merged recently) so we should update `RsKeywordMLKind` and the corresponding tests

ortem commented:
Rebased onto master (the first force-push) and fixed comments (the second force-push).

Also, I removed `"` symbols from `RsElementFeatureProviderTest` due to `[DANGEROUS_CHARACTERS] Name contains characters which can cause problems on Windows` warning.

Undin commented on build.gradle.kts:
```diff
@@ -450,6 +457,22 @@ project(":js") {
+project(":ml-completion") {
+    intellij {
+        // BACKCOMPAT: 2020.2
+        if (platformVersion >= 203) {
+            setPlugins(mlCompletionPlugin)
+        }
```
I suppose we also have to add dependency in `c-plugin` when we use `clion` as base IDE to avoid weird test fails because of `'navigation.class.hierarchy' is not found` error.
See `:toml` project as example

ortem replied:
Done

Undin reviewed (approved):
bors r+
